### DISCUSSION TITLE
Our Connection usage and XML vs SQL

### DISCUSSION THREAD

**Discussion by MikeNeilson** [opendcs/opendcs#263] [Development Practices]
@opendcs/opendcs-core-devs  @opendcs/opendcs-triage @wjonassen @krowvin @hoffmanjs @NathanHolcomb 
(Please inform anyone you've think I've left off of this as it will affect everyone.)

Alright, so in the process of fixing the RoutingScheduler for CWMS Usage, and trying to clean up the SQL DAO usage there's definitely going to be some problems with cleaning up the XML based DAOs in the same same way. Namely that's it XML and not SQL so odd to organize and handle the differences.

Long term, like in a version 8, I'd like to ditch the XML Database in favor of SQLite. There's absolutely still value in a simple on disk format, but that at least lets us Keep the actual data query mechanisms largely the same instead of split up.

In the mean time, I've found a project, https://calcite.apache.org/, that would allow us to wrap the current XML database behind a SQL layer, including having a JDBC Driver. This project allows us to pull the XML database tables into major, provide some specific meta data to indicate how it would operate as a :table" and takes care of the "SQL Magic" for us.

I'd like to do this for the following reason: I think we should get away from the concept of a "DatabaseConnectionOwner" and change that interface to a "DatabaseConnectionProvider" (See #222). Mostly to allow connection pooling to work correctly, and because we are very likely to be moving to a threaded world for more processes and you really aren't supposed to share java.sql.Connection objects between threads.

By wrapping the XML database behind a SQL/JDBC interface we can simplify the downstream code and keep that particular complexity exactly where it should be, at the IO layer. User configuration wouldn't need to change as the code can build the JDBC URL.

So this does add some complexity initially, but after that's done we can handle the pools more simply by say, just using HikariCP, and get away from the special case that is CWMS and how it handles connections.

It will also help with the migration to JDBI (instead of direct JDBC) access to the tables as we won't have "this type uses SQL, this type of db doesn't" issues. Also operating in environments like Tomcat where resources are control by the application server itself.

**Comment by agilmore2** [comment]
Sounds great. Would we still support import/export of XML? I've found value
(and pitfalls!) in a text viewable format, but that may be just old
fashioned me.

**Comment by MikeNeilson** [comment]
Yeah. I agree that the xml is still useful as an exchange/backup/archive format. Only thing I would want to change there is explicitly leaving off credentials.


Sent with BlackBerry Work (www.blackberry.com)
________________________________
From: agilmore2 ***@***.***>
Sent: Jul 27, 2023 9:41 AM
To: opendcs/opendcs ***@***.***>
Cc: Mike Neilson ***@***.***>; Author ***@***.***>
Subject: [URL Verdict: Neutral][Non-DoD Source] Re: [opendcs/opendcs] Our Connection usage and XML vs SQL (Discussion #263)


Sounds great. Would we still support import/export of XML? I've found value
(and pitfalls!) in a text viewable format, but that may be just old
fashioned me.

**Comment by MikeNeilson** [comment]
There's a few intermediate steps I have listed on paper, when I have time I'll add them to this discussion and create some milestones and projects for tracking.


Sent with BlackBerry Work (www.blackberry.com)
________________________________
From: agilmore2 ***@***.***>
Sent: Jul 27, 2023 9:41 AM
To: opendcs/opendcs ***@***.***>
Cc: Mike Neilson ***@***.***>; Author ***@***.***>
Subject: [URL Verdict: Neutral][Non-DoD Source] Re: [opendcs/opendcs] Our Connection usage and XML vs SQL (Discussion #263)


Sounds great. Would we still support import/export of XML? I've found value
(and pitfalls!) in a text viewable format, but that may be just old
fashioned me.

**Comment by MikeNeilson** [comment]
Okay, so this will roughly be the order I go in. Trying to lump by smallest PR possible.

1. Create an ServiceProvider interface (actually just repurposing "Database") to be responsible for all database creation (DECODES vs Timeseries). No user configuration change, just how the code deals with the user input.
2. ? Maybe tweak the Database interface scheme a little bit to include interfaces for concepts ("e.g. `... implements ScheduledEntry, Routing, etc` (this will make the test engine a little easier to design.)
3. Generate Schemas for the XML Database files. Expand to include "references"
4. Create the Apache Calcite modules (the XML schemas will make that much easier.)
5. Begin testing. This will start as a separate database implementation
   a. Create "opendcs-capability-engine" for junit5
7. Start working on the DatabaseConnectionOwner -> DatabaseConnectionProvider switch over.
6. Once testing is stable, remove old "XML implementation" (the bit from 5 just gets shifted silently in)
7. Begin change out to JDBI from direct JDBC access.


Some of that may not be in that order, will really depend on things, like I should probably really get the integration test engine working before a lot of this. But this is the basic list of what I think needs to happen.

**Comment by MikeNeilson** [comment]
A lot of this work has been started, additionally we've decided to replace the XML database with SQLite.
(XML will remain for import/export)
